### PR TITLE
feat: add swappable hasher, default to noble-hashes

### PR DESCRIPTION
**Pull request by wemeetagain**
**Motivation**

See #313 

We can replace as-sha256 with noble-hashes as long as we support our current interfaces and expose some mechanism to update the implementation for Lodestar's sake.

Its likely overkill to pass hash functions down thru our stack, and we can assume that a runtime will only require a single hash implementation.

**Description**

- Replace as-sha256 (and avoid loading any wasm) with noble-hashes sha256 implementation by default.
- Add a swapable "hasher" that can be updated before importing the persistent-merkle-tree / ssz or consumer libraries

Swapping the hasher with as-sha256, as will be required by lodestar, can be done like so:
```ts
// in lodestar, before anything else

import {hasher} from "@chainsafe/persistent-merkle-tree/hasher/as-sha256";
import {setHasher} from "@chainsafe/persistent-merkle-tree/hasher";

setHasher(hasher);

await import("./the-rest-of-the-application.js");
```

**Benchmarks**

Feel free to figure out how to speed this up if desired. I might be doing something stupid here.

```
yarn benchmark:files packages/persistent-merkle-tree/test/perf/hasher.test.ts

  hasher
    ✓ hash 2 Uint8Array 2250026 times - as-sha256                        0.4682371 ops/s    2.135670  s/op        -         12 runs   27.7 s
    ✓ hashTwoObjects 2250026 times - as-sha256                           0.4947299 ops/s    2.021305  s/op        -         12 runs   26.3 s
    ✓ hash 2 Uint8Array 2250026 times - noble                            0.1935327 ops/s    5.167086  s/op        -         11 runs   61.7 s
    ✓ hashTwoObjects 2250026 times - noble                               0.1418410 ops/s    7.050149  s/op        -          8 runs   63.4 s
```

### CODE REVIEW

File: packages/persistent-merkle-tree/src/hasher/noble.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import {sha256} from "@noble/hashes/sha256";
+import type {Hasher} from "./types";
+import {hashObjectToUint8Array, uint8ArrayToHashObject} from "./util";
+
+const digest64 = (a: Uint8Array, b: Uint8Array): Uint8Array => sha256.create().update(a).update(b).digest();
```
Comment (paulmillr) [inline packages/persistent-merkle-tree/src/hasher/noble.ts:5]: see if this helps:

```
const digest64 = (a: Uint8Array, b: Uint8Array): Uint8Array => sha256(concatBytes(a, b))
```

concatBytes can be imported from noble/hashes/utils

Comment (g11tech) [reply packages/persistent-merkle-tree/src/hasher/noble.ts:5]: would it give perf boost, from my understanding the underlying lib (for e.g. as-sha256) will be calling chunkified updates itself if you pass bigger input

**Review by g11tech — COMMENTED** [review]
looks amazing to me, leaving it unapproved if there is a followup on the perf improvement

File: .eslintrc.js [hunk 2]
```diff
@@ -81,6 +81,9 @@ module.exports = {
 
     // Prevents accidentally pushing a commit with .only in Mocha tests
     "no-only-tests/no-only-tests": "error",
+
+    // TEMP Disabled while eslint-plugin-import support ESM (Typescript does support it) https://github.com/import-js/eslint-plugin-import/issues/2170
```
Comment (mpetrunic) [inline .eslintrc.js:85]: Have you tried https://github.com/import-js/eslint-import-resolver-typescript ?

File: packages/persistent-merkle-tree/src/hasher/noble.ts (hunk 1 above)
Comment (paulmillr) [reply packages/persistent-merkle-tree/src/hasher/noble.ts:5]: probably won't, but just in case

**Comment by paulmillr** [comment]
I wonder why the perf difference here is 5x, while noble-hashes benchmarks only show 1.5x on 64b. Are you confident the benchmarks are proper?

**Comment by wemeetagain** [comment]
> I wonder why the perf difference here is 5x, while noble-hashes benchmarks only show 1.5x on 64b. Are you confident the benchmarks are proper?

We mostly hash two HashObjects, which is 3.5x. Not quite to 5x but in the ballpark.

**Comment by dapplion** [comment]
@wemeetagain can you set the benchmarks to use the WASM version? Else we go blind for Lodestar performance tracking

**Comment by dapplion** [comment]
Thanks @wemeetagain benchmarks look like before now :+1:

**Comment by paulmillr** [comment]
Could you merge this please? Current wasm breaks unsupported environments.

**Comment by wemeetagain** [comment]
> Could you merge this please?

Yes, looking for a :heavy_check_mark: from @ChainSafe/lodestar and will cut a release afterwards

**Review by dapplion — APPROVED** [review]

**Pull request merged**